### DISCUSSION TITLE
MorphOneThrough Relationship

### DISCUSSION THREAD

**Discussion by JeffreyDavidson** [laravel/framework#38944] [Ideas] [answered]
I know the MorphOneThrough relationship does not currently exist in the latest version of Laravel however I have a need for something like this. 

Has anyone else had a need for something like this and can share how they were able to handle this?

**Comment by staudenmeir** [comment]
What do the tables involved look like?

**Reply by JeffreyDavidson** [reply]
@staudenmeir This is the tables and models involved.

```
Titles
id
name

ModelA
id
name

ModelB
id
name

TitleChampionships
title_id
champion_id
champion_type
won_at
lost_at
```

**Comment by staudenmeir** [comment]
You can actually build this relationship yourself:

```php
class Title extends Model
{
    public function modelA()
    {
        return $this->hasOneThrough(ModelA::class, TitleChampionship::class, null, 'id', null, 'champion_id')
            ->where('champion_type', ModelA::class);
    }
}
```

**Reply by JeffreyDavidson** [reply]
@staudenmeir The only problem I have with that is that I would need it to be dynamic. Because there's no way of knowing what type it would be.

For example if a title has a current champion I would want it to grab the model that has a lost_at set to null but I wouldn't know if its going to return a ModelA or ModelB until it looks at the morphed entity.

```
$title->currentChampion. 
```

**Reply by JeffreyDavidson** [reply]
@staudenmeir Have you seen this situation before?

**Comment by staudenmeir** [comment, marked as answer]
Laravel doesn't support this kind of relationship because the SQL query would have to join a different table depending on the value of `champion_type` and that's not possible.

I created a package for complex cases like this that allows you to merge multiple relationships, but it requires a bit of setup:
https://github.com/staudenmeir/laravel-merged-relations

In your case, you would merge the two `HasOneThrough` relationships `modelA` and `modelB`. Also, the package currently doesn't offer a way to only return a single related result, only a collection. You would have to manually take the first collection item (e.g. with an accessor).

**Reply by cwilby** [reply]
> Laravel doesn't support this kind of relationship because the SQL query would have to join a different table depending on the value of champion_type and that's not possible.

Does this change if the consumer were to specify the possible values of `champion_type`?

See https://github.com/cwilby/framework/blob/bd17bfa3c861cd55be6c9ba8111ceebb3fb0bc7c/src/Illuminate/Database/Eloquent/Relations/MorphManyThrough.php#L74-L85

Then, you could get the table name for each model and iterate over each to add either a join or where exists:
https://github.com/cwilby/framework/blob/bd17bfa3c861cd55be6c9ba8111ceebb3fb0bc7c/src/Illuminate/Database/Eloquent/Relations/MorphManyThrough.php#L92-L106

**Reply by cwilby** [reply]
(Also `laravel-merged-relations` looks great for this, and it has solved a different problem of mine out of left field 👍 )

**Reply by JeffreyDavidson** [reply]
@staudenmeir What is your thoughts on this?

**Reply by staudenmeir** [reply]
> Does this change if the consumer were to specify the possible values of champion_type?

@cwilby In a single relationship?

**Reply by cwilby** [reply]
@staudenmeir If I understand you correctly - yes, the polymorphic type would need to be defined per relationship instance. 

A mechanic may work with many cars, where cars can morph to either one "Customer" or one "Dealership".

In that case, the mechanic model would define a "cars" relationship, where the `$throughParents` would be `[Customer::class, Dealership::class]`.

There is potential for an extra `select distinct owner_type from cars` to determine all the possible types, I imagine most will want some way to constrain which types are included in the relationship.

**Reply by staudenmeir** [reply]
@cwilby You can absolutely build your own custom relation class with this approach. Is that what you mean?

> Laravel doesn't support this kind of relationship

That might have been a bit ambiguous: I meant that you can't use a native Laravel relationship in this case.

**Reply by cwilby** [reply]
> @cwilby You can absolutely build your own custom relation class with this approach. Is that what you mean?
> 
> 
> 
> > Laravel doesn't support this kind of relationship
> 
> 
> 
> That might have been a bit ambiguous: I meant that you can't use a native Laravel relationship in this case. 
> 
> 

Exactly 👍

For context - Jeff and I are collaborating on either adding a new relation to Laravel; adding to has-many-deep; or creating a new package.

**Reply by edalzell** [reply]
@JeffreyDavidson @cwilby was there any result of this? I need a morphOneThrough right now, what did you end up on?

**Comment by Tklaversma** [comment]
Any updates on this? I need a morphOneThrough as well...
